### ISSUE TITLE
Decouple app model modules

### ISSUE THREAD

**Issue by IagoSRL** [loconomics/loconomics#421]
The set of modules that make up the 'app model' (data layer) are loaded all at start-up and attached into the `app.model` namespace.
This prevent us from splitting that code in bundles, where only the ones needed are loaded. As happened recently with the landing pages and will happen when we start to split our current big webapp bundle. Aside from other tooling benefits.

The base module for all this ones is /js/appmodel/AppModel.js, that requires all of them and provides initialization that is called at start-up in app.js.
The initialization takes care to set-up `localforage` and the rest client instance. Both are used by all modules. The first step is to refactor that. Maybe the rest client becomes its own module that can be required by each appmodel module without explicit initialization, delegated to the app startup.

With that solved, most modules can be easily decoupled from the namespace and must be explicitly required by each activity or viewmodel that use them.

A few modules may require some extra work (thinking on appmodel-account, userProfile and app.model.user entry) or being loaded all the time because are needed globally.

Too, the app.model methods `loadLocalCredentials` and `clearLocalData` needs to be refactored. In some way, all the modules must be able to register themselves or listen to a 'clearLocalData' event to perform clean-up when requested.
Maybe the 'AppModel' module, rather than removed, can be refactored to provide this general features and being required by the other modules.

**Comment by situatedbit** [comment]
Yes!

Even if we can refactor the typical app models this will be useful!

**Comment by IagoSRL** [comment]
@joshdanielson @situatedbit This is ready for testing, there are lot of internal changes and some complexities, I think I fixed every bug introduced in the refactoring but prefer to ask you for general testing of the app (if you can).

### Documentation of new organization

**Important** I removed the name 'App Model' and renamed the folder as '/data' where all 'data modules' lives (I think is more clear).

In subfolders goes closely related modules that don't provides the API to data. Are meant to be used only by the 'data modules' (files just under /data).
- /data/drivers: set-up of local storage, restClient for ajax calls. Are auto-initialized on first 'require'.
- /data/helpers: classes to assist creating the data modules, like the 'RemoteModel' or 'CachedData'. As a rule, if a module that 'seems' to belong here is used at other modules outside '/data', better to go at '/utils' rather than here.

The modules *appPresets, auth, credentials, session and "/data/drivers/" localforage, restClient* are new files from refactoring AppModel.js and AppModel-account.js, resulting in better separation of entities and responsibilities; they are fully documented with jsdocs, additional clarification code comments, and better organized to be easy to understand and use; I think are much more clear.

New is too the **/utils/SingleEvent.js** module: re-thinking events, moving away from 'nodejs/EventEmitter' for this uses in order to have events that are **declared** as properties in the classes/modules rather than rely on arbitrary strings and constants.
It's fully documented and has spec tests!

Most other data modules, for example /data/userProfile.js (from AppModel.userProfile.js) are adapted just the basic: removing the wrapping 'create' method, now are directly exported as modules, updated references and explicit access the rest client and the local storage, and the way of subscribe to the session 'clearLocalData' event (renamed as cacheCleaningRequested). The removal of some boilerplate and explicit references makes them looks more clear, but still this are the modules that require additional refactoring to prevent the commented problems with using shared data instances, managing state, the weird 'ModelVersion' (I tagged it as deprecated) and other details.

As special case, the userProfile module, widely used because contains the user name, user type and that; was previously used as 'app.model.userProfile.data', now I added next convention when requiring the module in the top:
`
var user = require('../data/userProfile').data;
` 
That way, can be used locally as `user.isServiceProfessional` and so; too, the old redundant observable `app.model.user()` was removed in favor of this (and simplified a bit more the code that used to use that).

-----------------

@situatedbit If you have time, I would appreciate your review of code and comments on this, but I understand if is not possible. Just the new modules, general organization, not the modules that are just adapted the 'basic' and skipping changes because of updated references.

**Comment by situatedbit** [comment]
I still need to test the code, but I've scanned over most of the code. Obviously I wasn't ready everything carefully. I like how the code is restructured as a result of this.

A couple small points:
- ``utils/ModelVersion.js``: why mark this as deprecated if you haven't replaced it yet? Seems like it should be marked as deprecated once there is a replacement that you want to move toward
- ``data/session.js``: I think the google analytics registration should be in its own module, and that module should subscribe to ``session`` events, and then register credentials with GA when the events are fired

**Comment by situatedbit** [comment]
``data/search.js`` introduces a bug in ``activities/searchJobTitle.js``. I'm not sure about your intentions in ``data/search.js``, so I did not try to fix it.

``search`` enforces one search at a time by aborting any pending searches when a new search is made. ``searchJobTitle`` initiates two searches sequentially: ``getJobTitle`` and then ``serviceProfessionalsByJobTitle``. There is a race condition here where ``search.serviceProfessionalsByJobTitle`` typically will abort the call just made to ``search.getJobTitle``. This results in the following layout on the ``searchJobTitle`` activity: 

![buggy](https://user-images.githubusercontent.com/6644016/27655335-e032baee-5bf9-11e7-8b64-fb43b9690eba.jpg)

instead of 

![not-buggy](https://user-images.githubusercontent.com/6644016/27655342-e56f7e2a-5bf9-11e7-80fc-43d03a6135a1.jpg)

**Comment by IagoSRL** [comment]
@situatedbit You are right about *ModelVersion*, was thinking in add the comment as a reminder to not extend its use; but we need and alternative approach first.

--------

About GA, yes looks better in its own module. Similar to this (register for the session event) I was thinking for 'sync userprofile' and 'resume onboarding' (now inline functions at app.js rather than methods on it's own data modules, see #504), but had the next doubt:
- If the session already opened/restored when attaching the handler, they never execute.
  - A public flag could help, but more boilerplate code
  - Offer a modification of the SingleEvent or Subscription returned by 'subscribe' to ask for execution if already the event was triggered almost one (a feature Promises have).
- The module still needs to be 'require'd to be loaded and being able to listen the event
  - We end with some 'require' lines at app.js without using the returned value, just to get them loaded and listening;
  - sometimes the order need to be ensured to prevent errors; not the case for GA though.

---------

Search data module: my mistake; that was an optimization to prevent too much requests but previously only affected the 'search.byTerm'; when refactoring this, I just though that 'could be fine' to not allow more than one search of any kind at same time, obviously a wrong assumption; I can update this logic to affect only each method, isolated.

Thanks to found this!!

**Comment by situatedbit** [comment]
@IagoSRL Re: GA, session, sync profile, etc.:

One way to address this is to convert ``app`` into a reusable module. I think you can implement this in small steps – you do not have to implement it all at once. Perhaps this is too complicated, but I like how it separates concerns. (``session`` does not have to know about Google Analytics, for example)

1. Create an ``App`` module. The module would include all of the ``requires`` that _any_ instance of the app would need.
2. Each bundle (a landing page, the dashboard "app", the public pages, etc) would initialize a new ``App``. Each bundle could have its own configuration for an ``App``. For example, a Landing Page configuration would not include Onboarding, but the dashboard version would include Onboarding.
3. In an initialization function, the ``App`` module would have a "boot" sequence that executed when the page loaded. The sequence could be implemented with ``SingleEvent``s. For example, an event handler for a ``beforeSessionRestore`` could ``require`` GA, and GA would then register for ``session`` events like ``restore`` and ``open``. Likewise, for the dashboard app, setting up onboarding and registering for session events could be added to the ``beforeSessionRestore`` event.

**Comment by IagoSRL** [comment]
I agree, I will separate GA as a module, makes sense, (and must include too the init code at app.js); can start putting it inside /data, but maybe I move it later because seems to not belong to there but to the app set-up.
Thinking incrementally, will start by putting the 'require' call to GA in the correct place at app.js and use the session events. I still consider your point 3, but that comes after next discussion/changes.

----------

I thought previously in the possibility of make app.js a class or module that can be *required* by others, but in the line of extract features from app.js, I'm thinking in extract a *startup* or *boot* module (other modules to extract can be the loading of globals/polyfills, native app specific set-up, shell-routing set-up, ...); not sure about organization but maybe an /js/app folder for them. Preventing to have again a module where 'everything fits' as happened with app.js.

This new modules would be required by the entry points/bundles we set-up: app.js, landingPage.js (webapp may become public.js, dashboard.js), so this ones are as short as possible. The idea is still fuzzy though.

Your approach is interesting, specially point 3. Maybe both things are valid, creating an App class that is responsible to load correctly some of that extracted modules, while others are optional loaded just by the bundles that need it.

Interesting discussion, we can continue at #500 or new issue. I want to keep moving to a more organized code base, easier to understand and update (specially the app set-up and initialization, had grow in a way that looks like a black box), and feedback is essential :-) Thanks!

**Comment by IagoSRL** [comment]
Aside from previous comment, @situatedbit your latest commit includes the removal of siteUrl set-up, I suppose was a mistake just wanting to force a test again *dev*.

**Comment by situatedbit** [comment]
Oops! My mistake. I've been doing that for 5 months without accidentally committing it. Fixed it.

**Comment by situatedbit** [comment]
I am unable to sign up as a new user. I submit the signup form, and the API call returns successfully, but I see this dialog: 

![error](https://user-images.githubusercontent.com/6644016/27744893-38602ecc-5d76-11e7-8974-33a1f0fdd23b.png)

Here's the api result ``http://dev.loconomics.com/api/v1/en-US/auth/signup?utm_source=web``: 
```json
{
  "redirectUrl": "/en-US/dashboard/welcome/",
  "userID": 827,
  "authKey": "u6gm3uJoROp3esLlpZKFkdiQLyAjlrReokuov6AwAfvjbprt3DC8sEQ9llqomZfhe6uCnpds9qsplwolfxBkmZKskw54Emrl3mnnJ8uNPvbxGwHiBKAFhISQpJQH6yCrmqSPuNFpCsKAuoEJPuReSsyBuDbIZqPs9hFytd4ilqrsR3aOlFAmJ6tGp4EmEd7R",
  "profile": {
    "userID": 827,
    "email": "matt+p04@loconomics.com",
    "firstName": "matt",
    "lastName": "pro4",
    "secondLastName": "",
    "businessName": null,
    "alternativeEmail": null,
    "phone": "6127091272",
    "canReceiveSms": false,
    "birthMonthDay": null,
    "birthMonth": null,
    "isServiceProfessional": false,
    "isClient": true,
    "isCollaborator": false,
    "onboardingStep": "welcome",
    "accountStatusID": 1,
    "createdDate": "2017-06-30T16:20:10.39-07:00",
    "updatedDate": "2017-06-30T16:20:10.39-07:00",
    "ownerStatusID": 0,
    "ownerAnniversaryDate": null,
    "photoUrl": "http://dev.loconomics.com/en-US/Profile/Photo/827?v=2017-06-30T16:20:10"
  },
  "onboardingStep": "welcome",
  "onboardingJobTitleID": 0
}
```

**Comment by situatedbit** [comment]
But, I _am_ authenticated during sign up. So I'm logged in, but I don't advance to the welcome screen.

**Comment by situatedbit** [comment]
I reproduced it for professionals. Same issue.

**Comment by IagoSRL** [comment]
@situatedbit I fixed and pushed a fix for last bug you commented (silly mistake, forgot 'return credentials').

**Comment by IagoSRL** [comment]
*Thinking out loud*

With this refactor we have a new data module named **credentials**, that stores the authorization keys and userID locally under the key name 'credentials' too (this key is in use from the beginning of the app, only that managed from the account/auth mixed code rather than in its own module).

With the upcoming renaming of user licenses/certifications as 'credentials' (will have a data module named 'userCredentials'), confusions may arise. Since the current module is about authentication credentials I'm thinking in rename it as **authCredentials** (and will appear next to current 'auth' module that provides the methods for authentication against the webservice). Will provide an auto migration of the local stored data (so users don't get logged out with the update).

**Comment by joshdanielson** [comment]
@IagoSRL another idea could be calling it userAuthKeys?

**Comment by IagoSRL** [comment]
Fixed the bug introduced at search, refactored Google Analytics code from session (some internal notes for further work after app.js full decoupling) and docs about ModelVersion, per conversations with @situatedbit.

**Comment by IagoSRL** [comment]
Finally, I renamed `data/credentials` as `data/userAuthKey` following @joshdanielson proposed name.

I fixed conflicts with *master*, it builds, some quick testing and works. Let me know if you find some bug that seems new, just in case was caused by this.

**Remember to review**
- new modules under /data
- NO more 'this.app.model...', now the data modules needs to be `required` at the top and used directly; the 'app.model' does not exist anymore.
- run `grunt clean build` after pull the updates.

**Additional note:** near all the API about how to use the data modules remains the same, will all the 'RemoteModel', 'ListRemoteModel', 'ModelVersion', shared Models and observables, and other classes and patterns still in use. As discussed with @situatedbit, this needs refactor too, and I was thinking in change how they works for things like to use plain data objects rather than Models for cache, retrieving a copy of the cached data rather than a shared reference (preventing the problems that make ModelVersion to appear --ideally this class would be removed at some point), and other practices both to get and use the data and to organize internally the modules. But that work was not started still and need to think well the pattern to apply.
